python/tgn.py

```diff
@@ -0,0 +1,695 @@
+
+        edge_feature = edge.properties.get("features", None)
+
+        if src_features is None:
+            # If node feature doesn't exist on one node, probably won't on any. Maybe this is wrong approach.
+            src_features = (
+                np.random.randint(
+                    0, 100, query_module_tgn.config[TGNParameters.NUM_NODE_FEATURES]
+                )
+                / 100
+            )
+        else:
+            assert type(src_features) is tuple
+            src_features = np.array(src_features)
+
+        if dest_features is None:
+            dest_features = (
+                np.random.randint(
+                    0, 100, query_module_tgn.config[TGNParameters.NUM_NODE_FEATURES]
+                )
+                / 100
+            )
+        else:
+            assert type(dest_features) is tuple
+            dest_features = np.array(dest_features)
+
+        if edge_feature is None:
+            edge_feature = (
+                np.random.randint(
+                    0, 100, query_module_tgn.config[TGNParameters.NUM_EDGE_FEATURES]
+                )
+                / 100
+            )
+        else:
+            assert type(edge_feature) is tuple
+            edge_feature = np.array(edge_feature)
+
+        src_features = torch.tensor(src_features, requires_grad=True, dtype=torch.float)
+        dest_features = torch.tensor(
+            dest_features, requires_grad=True, dtype=torch.float
+        )
+        edge_feature = torch.tensor(edge_feature, requires_grad=True, dtype=torch.float)
+
+        query_module_tgn_batch.node_features[src_id] = src_features
+        query_module_tgn_batch.node_features[dest_id] = dest_features
+
+        query_module_tgn_batch.edge_features[edge_idx] = edge_feature
+
+        query_module_tgn_batch.sources = np.append(
+            query_module_tgn_batch.sources, src_id
+        )
+        query_module_tgn_batch.destinations = np.append(
+            query_module_tgn_batch.destinations, dest_id
+        )
+        query_module_tgn_batch.timestamps = np.append(
+            query_module_tgn_batch.timestamps, timestamp
+        )
+        query_module_tgn_batch.edge_idxs = np.append(
+            query_module_tgn_batch.edge_idxs, edge_idx
+        )
+        query_module_tgn_batch.labels = np.append(
+            query_module_tgn_batch.labels, np.array([[src_label, dest_label]]), axis=0
+        )
+
+
+def reset_tgn_batch(batch_size: int):
+    global query_module_tgn_batch
+    query_module_tgn_batch = QueryModuleTGNBatch(
+        0,
+        np.empty((0, 1), dtype=int),
+        np.empty((0, 1), dtype=int),
+        np.empty((0, 1), dtype=int),
+        np.empty((0, 1), dtype=int),
+        {},
+        {},
+        batch_size,
+        np.empty((0, 2), dtype=int),
+    )
+
+
+def reset_tgn():
+    global all_embeddings, query_module_tgn_batch, query_module_tgn, all_edges
+
+    all_embeddings = {}
+    reset_tgn_batch(0)
+    all_edges = set()
```

Comment:
Why is this in the module?

Reply:
Embedding and such inference variables are more part of the module than core TGN, that's why I put them in the module. It seems to me such variables belong in TGN. What do you think?